### ISSUE TITLE
Tracking buttons with the same name

### ISSUE THREAD

**Issue by 444B** [444B/streamlit-analytics2#26]
Original issue by amirmk89 on 2023-01-12 15:10:09+00:00

Hi, awesome tool!

I have a use case where several buttons have the same text, and currently the analytics tool tracks all of them as a single button. Each of the buttons has a unique key, and I thought it could be useful to support key based indexing instead of, e.g., button label based.
Is this possible in streamlit-analytics?

Thanks!

**Comment by 444B** [comment]
Comment by nathanjones4323 on 2023-03-21 15:53:51+00:00

Same here

I believe the issue happens during the "monkey patching" of the widgets inside `main.py`
[Here](https://github.com/jrieke/streamlit-analytics/blob/main/streamlit_analytics/main.py#L34-L64)
&
[Here](https://github.com/jrieke/streamlit-analytics/blob/main/streamlit_analytics/main.py#L87-L226)
&
[Here](https://github.com/jrieke/streamlit-analytics/blob/main/streamlit_analytics/main.py#L278-L307)

The widget state/key gets set by default to the label of the widget during the "monkey patching", so even if you have a key set for the widget inside of your app, the `state_dict` used to track widget interaction will use the label of the widget.

Maybe you could move the "monkey patching" inside of your app page, but that seems messy. Ideally, we have the "monkey patched" widgets inside `main.py` inherit the key from the app file.

**Comment by 444B** [comment]
@amirmk89 @nathanjones4323  is this still relevant?

**Comment by emigre459** [comment]
+1 vote from me on this one. I was about to submit the feature request in fact! Right now I have a use case in which I'm tracking buttons I generate as part of the answer from a large language model (LLM) so the user has the ability to select some pre-made responses/followups in the conversation. Since the labels (the followup question) is pretty much different every single time the buttons are generated, using labels to track the clicks is messy. Much easier to track "did the user click on a followup" by using the key param (and that also helps us keep what is tracked separate from what the user sees in the UI, which seems like a good separation of concerns).

**Comment by 444B** [comment]
@emigre459 Hi! Nice to see you again  
I will focus on this fix/feature for the next release  
Can you possible provide me with some code to replicate the issue?
